File: app/src/main/java/com/tibadev/alimansour/prophetstories/MainActivity.java
package com.tibadev.alimansour.prophetstories;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.ShareActionProvider;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.startapp.android.publish.StartAppAd;
import com.startapp.android.publish.StartAppSDK;
import com.startapp.android.publish.model.AutoInterstitialPreferences;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutionException;

public class MainActivity extends AppCompatActivity {
    ListView listView;
    List<Story> stories = null;
    String strProphet = null;
    String strContent = null;
    NetworkCheck networkCheck;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        try {
            // Check Device Connectivity & view Ads if it is connected:
            NetworkCheck networkCheck = new NetworkCheck();
            if (networkCheck.check(this)) {
                // Check for updates
                String latestVersion = "";
                String currentVersion = getCurrentVersion();
                try {
                    latestVersion = new GetLatestVersion().execute().get();
                } catch (InterruptedException | ExecutionException e) {
                    e.printStackTrace();
                }

                // Show Update Alert Dialog at startup if versions are not equal
                if (!currentVersion.equals(latestVersion)) {
                    final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    builder.setIcon(R.drawable.ic_launcher);
                    builder.setTitle(R.string.text_update_alert_title);
                    builder.setMessage(R.string.text_update_alert_message);
                    builder.setPositiveButton(R.string.btn_update_alert_update, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            //Click button action
                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=com.tibadev.alimansour.prophetstories")));
                            dialog.dismiss();
                        }
                    });
                    builder.setCancelable(false);
                    builder.show();
                }
                // View Ads
                StartAppSDK.init(this, "210823214", false);
                StartAppAd.disableSplash();
                StartAppAd.enableAutoInterstitial();
                StartAppAd.setAutoInterstitialPreferences(new AutoInterstitialPreferences().setActivitiesBetweenAds(5));
            }

            // Check for permissions
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {

                    // Check for Access fine location permission
                    if (checkSelfPermission(android.Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                        // Permission is available, start the app
                    } else {
                        requestPermissions(new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION}, 0);
                    }

                    // Should we show an explanation?
                    if (shouldShowRequestPermissionRationale(android.Manifest.permission.ACCESS_FINE_LOCATION)) {
                        // Explain to the user why we need ACCESS_FINE_LOCATION
                        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                        builder.setIcon(R.drawable.ic_my_location_black_24dp);
                        builder.setTitle("استخدام تحديد المواقع");
                        builder.setMessage("التطبيق يقوم باستخدام صلاحية تحديد المواقع في هاتفك لعرض إعلانات مناسبة لك.");
                        builder.setPositiveButton("إخفاء", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                //Click button action
                                dialog.dismiss();
                            }
                        });
                        builder.setCancelable(true);
                        builder.show();
                    }
                }
            } catch (Exception EX) {
            }

            listView = (ListView) findViewById(R.id.list);
            try {
                XMLPullParserHandler parser = new XMLPullParserHandler();
                stories = parser.parse(getAssets().open("stories.xml"));
                listView.setAdapter(new MyAdapter(this, android.R.layout.simple_list_item_1, stories));
            } catch (IOException e) {
                e.printStackTrace();
            }
            final List<Story> finalStories = stories;
            listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                    // Retrieve story data
                    strProphet = finalStories.get(position).getProphet();
                    strContent = finalStories.get(position).getContent();
                    showStory();
                }
            });
        } catch (Exception e) {
        }
    }

    private void showStory() {
        Intent intent = new Intent(MainActivity.this, StoryActivity.class);
        intent.putExtra("prophet", strProphet);
        intent.putExtra("content", strContent);
        startActivity(intent);
    }

    public class MyAdapter extends ArrayAdapter<Story> {


        MyAdapter(Context context, int resource, List<Story> objects) {
            super(context, resource, objects);
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            LayoutInflater inflater = (LayoutInflater) getSystemService(LAYOUT_INFLATER_SERVICE);
            View view = inflater.inflate(R.layout.list_item, parent, false);
            TextView txtItemCompanion = (TextView) view.findViewById(R.id.txt_item_prophet);
            Story story = stories.get(position);
            txtItemCompanion.setText(story.getProphet());
            return view;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        try {
            if (requestCode == 0) {
                // Request for call phone permission.
                if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
                    // Permission request was denied.
                    Toast.makeText(this, R.string.text_location_permission_denied, Toast.LENGTH_SHORT).show();
                }
            }
        } catch (Exception EX) {
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        // Retrieve the share menu item
        MenuItem shareItem = menu.findItem(R.id.action_share);

        // Now get the ShareActionProvider from the item
        ShareActionProvider mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(shareItem);
        mShareActionProvider.setShareIntent(doShare());

        MenuItem rateItem = menu.findItem(R.id.action_rate);
        rateItem.setOnMenuItemClickListener(new MenuItem.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                //Try Google play
                intent.setData(Uri.parse("market://details?id=com.tibadev.alimansour.prophetstories"));
                if (!MyStartActivity(intent)) {
                    //Market (Google play) app seems not installed, let's try to open a webbrowser
                    intent.setData(Uri.parse("https://play.google.com/store/apps/details?com.tibadev.alimansour.prophetstories"));
                    if (!MyStartActivity(intent)) {
                        //Well if this also fails, we have run out of options, inform the user.
                        Toast.makeText(MainActivity.this, "عفواً لا يمكن فتح تطبيق جوجل بلاي ، برجاء تثبيت التطبيق.", Toast.LENGTH_SHORT).show();
                    }
                }
                return false;
            }
        });
        return true;
    }

    // Check for updates
    private String getCurrentVersion() {
        PackageManager pm = this.getPackageManager();
        PackageInfo pInfo = null;

        try {
            pInfo = pm.getPackageInfo(this.getPackageName(), 0);

        } catch (PackageManager.NameNotFoundException e1) {
            e1.printStackTrace();
        }

        assert pInfo != null;
        return pInfo.versionName;
    }

    public Intent doShare() {
        // populate the share intent with data
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_SUBJECT, "تطبيق قصص الأنبياء - أندرويد");
        intent.putExtra(Intent.EXTRA_TEXT, "قصص الأنبياء - أندرويد\nطيبة ديف للبرمجيات\n\nhttps://play.google.com/store/apps/details?com.tibadev.alimansour.prophetstories");
        return intent;
    }

    private boolean MyStartActivity(Intent aIntent) {
        try {
            startActivity(aIntent);
            return true;
        } catch (ActivityNotFoundException e) {
            return false;
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //ToDo: noinspection SimplifiableIfStatement
        switch (id) {
            case R.id.action_our_apps:
                Intent intent = new Intent(Intent.ACTION_VIEW);
                //Try Google play
                intent.setData(Uri.parse("market://dev?id=8256417273842663752"));
                if (!MyStartActivity(intent)) {
                    //Market (Google play) app seems not installed, let's try to open a webbrowser
                    intent.setData(Uri.parse("https://play.google.com/store/apps/dev?id=8256417273842663752"));
                    if (!MyStartActivity(intent)) {
                        //Well if this also fails, we have run out of options, inform the user.
                        Toast.makeText(MainActivity.this, "عفواً لا يمكن فتح تطبيق جوجل بلاي ، برجاء تثبيت التطبيق.", Toast.LENGTH_SHORT).show();
                    }
                }
                break;
            case R.id.action_settings:

                break;
            case R.id.action_about:
                Intent intentabout = new Intent(MainActivity.this, AboutActivity.class);
                startActivity(intentabout);
                break;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/tibadev/alimansour/prophetstories/activities/MainActivity.kt
package com.tibadev.alimansour.prophetstories.activities

import android.app.Dialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.ShareActionProvider
import androidx.appcompat.widget.Toolbar
import androidx.core.view.MenuItemCompat
import com.tibadev.alimansour.prophetstories.BuildConfig
import com.tibadev.alimansour.prophetstories.R
import com.tibadev.alimansour.prophetstories.models.Story
import com.tibadev.alimansour.prophetstories.util.Network
import com.tibadev.alimansour.prophetstories.util.XMLPullParserHandler
import com.tibadev.alimansour.prophetstories.util.showInterstitialAd
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.jsoup.Jsoup
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private lateinit var listView: ListView
    private lateinit var stories: List<Story>
    private lateinit var strProphet: String
    private lateinit var strContent: String
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        try {
            val network = Network(this)
            if (network.isConnected()) checkForUpdates()

            listView = findViewById<View>(R.id.list) as ListView
            try {
                val parser = XMLPullParserHandler()
                stories = parser.parse(assets.open("stories.xml"))
                listView.adapter = MyAdapter(this, android.R.layout.simple_list_item_1, stories)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            val finalStories = stories
            listView.onItemClickListener =
                OnItemClickListener { _, _, position, _ ->
                    strProphet = finalStories[position].prophet
                    strContent = finalStories[position].content
                    showInterstitialAd(this)
                    showStory()
                }
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, e.message.toString())
        }
    }

    private fun showStory() {
        val intent = Intent(this@MainActivity, StoryActivity::class.java)
        intent.putExtra("prophet", strProphet)
        intent.putExtra("content", strContent)
        startActivity(intent)
    }

    inner class MyAdapter internal constructor(
        context: Context?,
        resource: Int,
        objects: List<Story>?
    ) : ArrayAdapter<Story?>(
        context!!, resource, objects!!
    ) {
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val inflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view = inflater.inflate(R.layout.list_item, parent, false)
            val txtItemCompanion = view.findViewById<View>(R.id.txt_item_prophet) as TextView
            val (prophet) = stories[position]
            txtItemCompanion.text = prophet
            return view
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        // Retrieve the share menu item
        val shareItem = menu.findItem(R.id.action_share)

        // Now get the ShareActionProvider from the item
        val mShareActionProvider =
            MenuItemCompat.getActionProvider(shareItem) as ShareActionProvider
        mShareActionProvider.setShareIntent(doShare())
        val rateItem = menu.findItem(R.id.action_rate)
        rateItem.setOnMenuItemClickListener {
            val intent = Intent(Intent.ACTION_VIEW)
            //Try Google play
            intent.data = Uri.parse("market://details?id=com.tibadev.alimansour.prophetstories")
            if (!myStartActivity(intent)) {
                //Market (Google play) app seems not installed, let's try to open a webbrowser
                intent.data =
                    Uri.parse("https://play.google.com/store/apps/details?com.tibadev.alimansour.prophetstories")
                if (!myStartActivity(intent)) {
                    //Well if this also fails, we have run out of options, inform the user.
                    Toast.makeText(
                        this@MainActivity,
                        "عفواً لا يمكن فتح تطبيق جوجل بلاي ، برجاء تثبيت التطبيق.",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
            false
        }
        return true
    }

    private fun doShare(): Intent {
        // populate the share intent with data
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_SUBJECT, "تطبيق قصص الأنبياء - أندرويد")
        intent.putExtra(
            Intent.EXTRA_TEXT,
            "قصص الأنبياء - أندرويد\nطيبة ديف للبرمجيات\n\nhttps://play.google.com/store/apps/details?com.tibadev.alimansour.prophetstories"
        )
        return intent
    }

    private fun myStartActivity(aIntent: Intent): Boolean {
        return try {
            startActivity(aIntent)
            true
        } catch (e: ActivityNotFoundException) {
            false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_privacy_policy ->
                Network(this).openLink(getString(R.string.privacy_policy_url))

            R.id.action_our_apps -> showOurApps()

            R.id.action_about -> {
                val intent = Intent(this@MainActivity, AboutActivity::class.java)
                startActivity(intent)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showOurApps() {
        try {
            startActivity(
                Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse(getString(R.string.our_apps_market_url))
                )
            )
        } catch (exception: ActivityNotFoundException) {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.our_apps_url))))
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun checkForUpdates() {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                val currentVersion = BuildConfig.VERSION_NAME
                val latestVersion = CoroutineScope(Dispatchers.IO).async {
                    try {
                        // retrieve the latest version from play store at runtime
                        val urlOfAppFromPlayStore =
                            applicationContext.getString(R.string.app_url) + applicationContext
                                .packageName
                        val doc = Jsoup.connect(urlOfAppFromPlayStore).timeout(30000).get()
                        if (doc != null) {
                            val element = doc.getElementsContainingOwnText("Current Version")
                            for (ele in element) {
                                if (ele.siblingElements() != null) {
                                    val siblingElements = ele.siblingElements()
                                    for (sibElement in siblingElements) {
                                        return@async sibElement.text()
                                    }
                                }
                            }
                        } else {
                            return@async ""
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }.await()

                // View Update prompt if the installed version is not the latest published one
                if (currentVersion != latestVersion) {
                    val mDialog = Dialog(this@MainActivity)
                    mDialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
                    mDialog.setContentView(R.layout.custom_alert_dialog)
                    mDialog.setCancelable(false)
                    val titleTextView = mDialog.findViewById<TextView>(R.id.title_text_view)
                    val messageTextView = mDialog.findViewById<TextView>(R.id.message_text_view)
                    val applyButton = mDialog.findViewById<Button>(R.id.apply_button)
                    val cancelButton = mDialog.findViewById<Button>(R.id.cancel_button)
                    titleTextView.text = getString(R.string.update_alert_title)
                    messageTextView.text = getString(R.string.update_alert_message)
                    applyButton.text = getString(R.string.update_alert_update_button)
                    cancelButton.text = getString(R.string.alert_later_button)
                    applyButton.setOnClickListener {
                        startActivity(
                            Intent(
                                Intent.ACTION_VIEW, Uri.parse(
                                    applicationContext.getString(R.string.app_market_url) + applicationContext.packageName
                                )
                            )
                        )
                        mDialog.dismiss()
                    }
                    cancelButton.setOnClickListener { mDialog.dismiss() }
                    mDialog.show()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}